### DISCUSSION TITLE
JSConfig / TSConfig

### DISCUSSION THREAD

**Discussion by Raynos** [voxpelli/types-in-js#2] [Best Practices / Patterns]
To get the most out of leveraging jsdoc you want to configure your project.

In my experience, I found that a `jsconfig.json` file behaves better then a `tsconfig.json` file.

I have an example [jsconfig.json](https://github.com/Raynos/fake-s3/blob/master/jsconfig.json) file

```json5
{
  "extends": "./types/base-tsconfig.json",
  "compilerOptions": {
    "baseUrl": "./",
    "paths": {
      "@pre-bundled/tape": ["./types/pre-bundled__tape"],
      "@pre-bundled/rimraf": ["./types/pre-bundled__rimraf"],
      "*" : ["./types/*"]
    }
  },
  "include": [
    "types/**/*.d.ts",
    "node_modules/tape-harness/index.js",
    "index.js",
    "test/**/*.js"
  ]
  // "exclude": ["node_modules"]
}
```

For my personal sanity I choose to put some defaults in another file. I copy that file into basically every project and it's very similar and not project specific ( https://github.com/Raynos/fake-s3/blob/master/types/base-tsconfig.json ).

The content of the `jsconfig.json` is project specific and it states the files and dependencies for this project.

**Comment by voxpelli** [comment]
I for a long while didn't know about `jsconfig.json`, so I've stuck with `tsconfig.json` as the only difference except for the name that I know of is that `jsconfig.json` implies `allowJs: true`.

My typical `tsconfig.json` right now looks like:

```json5
{
  "files": [
    "index.js"
  ],
  "include": [
    "lib/**/*.js",
    "test/**/*.js"
  ],
  "compilerOptions": {
    /* Basic Options */
    "target": "ES2017",
    "lib": ["ESNext.Array"],
    "module": "commonjs",
    "allowJs": true,
    "checkJs": true,
    "resolveJsonModule": true,
    "noEmit": true,
    "esModuleInterop": true,

    "removeComments": true,

    /* Strict Type-Checking Options */
    "strict": true,
    "noImplicitThis": true,
    "noImplicitAny": true,
    "strictNullChecks": true,

    /* Additional Checks */
    "noUnusedLocals": true,
    "noUnusedParameters": true,
    "noImplicitReturns": true
  }
}
```

I will eventually move to the setup of eg. @sindresorhus, with an extracted config that's shared between all of my projects: https://github.com/sindresorhus/tsconfig

I have so far felt that I'm still in a state of iteration and that such an abstraction for me would still be a bit premature. But eventually I will make a `@voxpelli/tsconfig` module.

**Reply by voxpelli** [reply]
I have now moved to such an extracted config, and I opted to do so similar to how TypeScripts own base configs are: https://github.com/tsconfig/bases

So I eg. have both a `node14.json` and a `node12.json`: https://github.com/voxpelli/tsconfig

**Reply by Raynos** [reply]
Do you have a usage examples ?

**Reply by voxpelli** [reply]
Here’s one: https://github.com/voxpelli/node-pg-pubsub/blob/main/tsconfig.json

**Comment by Raynos** [comment]
>  so I've stuck with tsconfig.json as the only difference except for the name that I know of is that jsconfig.json implies allowJs: true.

I've found that there are subtle differences in way the text editor behaves in the existance of a tsconfig / jsconfig file. The existance of a tsconfig file with no typescript files can make the editor behave less javascript friendly.

**Reply by voxpelli** [reply]
Interesting, any particular ways you can remember where it behaves less friendly? Maybe I should make the switch.

**Reply by Raynos** [reply]
Go to definition behaves differently based on jsconfig / tsconfig. The existance of a tsconfig.json breaks go to definition.

https://github.com/microsoft/TypeScript/issues/40978

**Reply by voxpelli** [reply]
Worth mentioning in this context: https://github.com/voxpelli/types-in-js/discussions/6#discussioncomment-218820

@Raynos Thoughts on `jsconfig` in relation to that?

**Comment by voxpelli** [comment]
Seems like `jsonfig.json` nowadays implies much more than just the `allowJs: true` that the VSCode docs implies: https://github.com/eslint/eslint/issues/18100#issuecomment-1971500684

**Reply by voxpelli** [reply]
Here's a list of all additions on top of `allowJs: true` that has happened:

* **Aug 25, 2016:** When `maxNodeModuleJsDepth` was changed to `0` as default for TS as a whole, it was kept as `2` for `jsconfig.json` (without updating the docs it seems) https://github.com/microsoft/TypeScript/commit/c0309fa78e5dbf7dd699cf21ea59ed9cb24268ad / https://github.com/microsoft/TypeScript/pull/10538
* **Sep 23, 2016:** `allowSyntheticDefaultImports: true` was added in https://github.com/microsoft/TypeScript/commit/12578a5cc6f2d9542bea1f3f47c8475d8c1d129a / https://github.com/microsoft/TypeScript/pull/11070
* **Oct 12, 2016:** `skipLibCheck: true` was added in https://github.com/microsoft/TypeScript/commit/460de66311b26457481a6d2055231e00fa24e705 / https://github.com/microsoft/TypeScript/pull/11399
* **Jan 25, 2018:** ` noEmit: true` was added in https://github.com/microsoft/TypeScript/commit/4af47f188772591ecd996eb18a853c2cf1cd83f2 / https://github.com/microsoft/TypeScript/pull/21505

And that's where we [remain today](https://github.com/microsoft/TypeScript/blob/2d70b57df4b64a3daef252abb014562e6ccc8f3c/src/compiler/commandLineParser.ts#L3433) with `jsconfig.json` implying:

```json
{
  "allowJs": true,
  "maxNodeModuleJsDepth": 2,
  "allowSyntheticDefaultImports": true,
  "skipLibCheck": true,
  "noEmit": true
}
```

But the only documentation of what it implies that I can find is the [one in VSCode docs](https://code.visualstudio.com/docs/languages/jsconfig), and those say:

> **Tip:** `jsconfig.json` is a descendant of [tsconfig.json](https://www.typescriptlang.org/docs/handbook/tsconfig-json.html), which is a configuration file for TypeScript. `jsconfig.json` is `tsconfig.json` with `"allowJs"` attribute set to `true`.

It seems long overdue for an update: https://github.com/microsoft/vscode-docs/blame/91835d57149c4352073c22300487a18baeefe13c/docs/languages/jsconfig.md#L18

(Which I will do following this comment)

**Reply by voxpelli** [reply]
Separated into a new post: https://github.com/voxpelli/types-in-js/discussions/25

TLDR is essentially: Don't use `jsconfig.json`, its an oddity that sets weird options.
